[WIP] Update 2.5 migration guide (#612)
ezsystems/developer-documentation: docs/migrating/migrating_from_ez_publish_platform.md

eZ Publish Platform (5.x) was a transitional version of the eZ CMS, bridging the gap between the earlier generation called eZ Publish (sometimes referred to as *legacy*), and the current eZ Platform and eZ Platform Enterprise Edition for developers.

eZ Publish Platform (5.x) introduced a new Symfony-based technology stack that could be run along the old (*legacy*) one. This bridging is still possible using something called Legacy Bridge, an optional package for eZ Platform. This fluid change allows eZ Publish users to migrate to eZ Platform gradually, using the bridging as an intermediary stepping stone.

## Upgrade process

An upgrade from eZ Publish Platform 5.4.x (Enterprise edition) or 2014.11 (Community edition) to newer versions of eZ Platform must be performed in stages. It can be done in many different ways, but a recommended approach is to:

> 1. _If you need to migrate ezflow data:_ Migrate from eZ Publish Platform directly to the v1.7LTS release _(covered on this page)_.

Author: NB: The fact that this package needs to be done on 1.7 does not make much sense given legacy bridge is not supported on 1.7 to begin with.

    - Optional; Add legacy bridge to be able to gradual migrate your legacy code if you have it
    - Once ready; Migrate eZ FLow to eZ Landing page

> 2. You can then proceed with consecutive upgrades to further versions, covered on [v1.7 LTS an 2.5LTS](releases/updating_ez_platform.md).

Reviewer: ```suggestion
2. You can then proceed with consecutive upgrades to further versions, covered on [v1.7 LTS and 2.5 LTS](releases/updating_ez_platform.md).
```

    2. "Hybrid setup" using Legacy Bridge is a supported option for 1.13LTS and 2.5LTS series. This means you can plan for a more gradual migration if you want, just like you could on eZ Publish Platform 5.x, with a more feature-mature version of eZ Platform, Symfony and even Legacy bridge itself. This is a great option for those who want the latest features. The downside is a more complex setup to develop and maintian, given you run two systems in paralell, and the overall migration might take much longer doing using an iterative/gradual approach. 

    3. Additionally there are some other topics to be aware of for the code migration from eZ Publish to eZ Platform:

        - [Field Types reference](../api/field_type_reference.md) for overview of Field Types that do and don't exist in eZ Platform

>         - eZ Platform RichText Field Type capabilities: You'll need v2.5.1 or higher for custom attributes migration support.

Reviewer: I think it's good to mention that you need 2.5.1 if you want custom attributes. But if you do `composer update` you will automatically have the latest release, right?
I don't see why people will stay on 2.5.0 during the migration.

Author: You are right, that comment is not longer needed (it was written at the time 2.5.1 came out).

Author: ```suggestion
        - eZ Platform RichText Field Type capabilities: You'll need latests v2.5LTS for custom attributes migration support.
```

`composer require --no-update "vendor/package:~1.3.0"`

Adapt the command with your `vendor`, `package`, version number, and add `"–dev"` if a given package is for dev use only. Also check if there are other changes in `composer.json` you should move over.

###### 2.2.2 Install XmlText Field Type

While no longer bundled, the XmlText Field Type still exists and is needed to perform a migration from eZ Publish's XmlText to the new docbook-based format used by the RichText Field Type. If you plan to use Legacy Bridge for a while before migrating content, you'll also need this for rendering content with XMLText. From `<new-ez-root>` execute:

> `composer require --no-update "ezsystems/ezplatform-xmltext-fieldtype:^1.7.3"`

Author: Blocker: This version does not support 1.7LTS: https://github.com/ezsystems/ezplatform-xmltext-fieldtype/blob/v1.7.3/composer.json#L13

!!! note

>     Be aware this XmlText Field Type in eZ Platform uses the Content View system introduced in eZ Platform 1.0, so make sure you adapt custom templates and override rules if you plan to use this for rendering xmltext content _(in Legacy Bridge setup)_.

Reviewer: ```suggestion
    Be aware this XmlText Field Type in eZ Platform uses the Content View system introduced in eZ Platform v1.0, so make sure you adapt custom templates and override rules if you plan to use this for rendering xmltext content _(in Legacy Bridge setup)_.
```

##### 3.1. Execute update SQL

Import to your database the changes provided in one of the following files. It's also recommended reading inline comments as you might not need to run some queries *(if you for instance upgrade from 5.4.11)*:

Postgres:
- `<new-ez-root>/vendor/ezsystems/ezpublish-kernel/data/update/postgres/dbupdate-5.4.0-to-6.13.0.sql`
- `<new-ez-root>/vendor/ezsystems/ezpublish-kernel/data/update/postgres/dbupdate-6.13.0-to-7.5.7.sql`

MySQL/MariaDB:
- `<new-ez-root>/vendor/ezsystems/ezpublish-kernel/data/update/mysql/dbupdate-5.4.0-to-6.13.0.sql`
- `<new-ez-root>/vendor/ezsystems/ezpublish-kernel/data/update/mysql/dbupdate-6.13.0-to-7.5.7.sql`

> !!! note  "Change to UTF8mb4 for MySQL/MariaDB"

Reviewer 2: This is not the only thing needed.

I think we should just instruct user to run all upgrade things mentioned at https://doc.ezplatform.com/en/latest/updating/4_update_database/, but skip [LandingPage -> Page builder upgrade](https://doc.ezplatform.com/en/latest/updating/4_update_2.2/#page-builder) mentioned for 2.2

Then continue here afterwards

Author: So I'm trying to provide a fast pass here as many steps there are not relevant when coming directly from 5.4.

We might be missing instructions for creating form container thought: https://doc.ezplatform.com/en/latest/updating/4_update_2.3/#form-builder

Can add that in `3.2.3 Add other eZ Enterprise schemas (eZ Enterprise only)`

Reviewer 2: But why should you skip all of that?
You for sure need to upgrade db schema to 2.5 ( or no migration scripts will run at all). But  you should  for instance also create new top folder for "Forms" etc. and everything else explained there in order to get a fully functional 2.5

Reviewer 2: But why should you skip all of that?
You for sure need to upgrade db schema to 2.5 ( or no migration scripts will run at all). But  you should  for instance also create new section for "Forms" (2.4) etc. and everything else explained there in order to get a fully functional 2.5.

Author: > You for sure need to upgrade db schema to 2.5 ( or no migration scripts will run at all)

There is:
```
MySQL/MariaDB:
- `<new-ez-root>/vendor/ezsystems/ezpublish-kernel/data/update/mysql/dbupdate-5.4.0-to-6.13.0.sql`
- `<new-ez-root>/vendor/ezsystems/ezpublish-kernel/data/update/mysql/dbupdate-6.13.0-to-7.5.7.sql`
```

> But you should for instance also create new section for "Forms" (2.4) etc. and everything else explained there in order to get a fully functional 2.5.

Yes I'm missing that one, as mentioned above, there is also a few other things we should cover.

Otherwise I think I cover everything else relevant when coming from 5.4, so to go true all of it then:
- KO: `<1.7 Changes to Solr Bundle` => AFAIK invalid, there are more changes to schema then what is mentioned
    - So we should add own section here on using the new solr schema instead
- OK: `<1.13 Changes to permissions` => Covered by dbupdate-5.4.0-to-6.13.0.sql AFAIK
- OK: `<1.13 Changes to form-uploaded files` => Irrelevant, is about LandingPage
- OK: `<1.13 ezsearch_return_count table removal` => This was covered in 5.4, we could add it again if we want
- OK: `<1.13 Increased password hash length` => Should be covered by dbupdate-5.4.0-to-6.13.0.sql AFAIK
- OK: `<2.2 Change from UTF8 to UTF8MB4` => Covered in the links here
- OK: `<2.2 Page builder` => Not relevant as we install the table clean with all fixes in step to install EE tables here
- OK: `<2.2 Migrate Landing Pages` => Covered below here
- OK: `<2.3 Changes to timestamp` => Covered by the NEW (PR) `dbupdate-6.13.0-to-7.5.7`
- KO: `<2.3 Form builder` => We should add that step here or link to it, as described in comment above.
- OK: `<2.4 Workflow` => Covered by instructions to install all EE tables
- ??: `<2.4 Changes to the Forms folder` => Kind of assume command above to create folder uses right section, but worth checking
- OK: `<2.4 Changes to Custom tags` => Irrelevant, install does not yet use RichText

Then there is 2.5 page:
- OK: All DB steps are covered by `dbupdate-6.13.0-to-7.5.7`
- OK: `DoctrineSchemaBundle ` is already enabled as we instruct people to use clean 2.5.x as starting point here
- OK: matrix steps is covered here
- ??: Cache flush can be relevant, even if not for the same reasons, so we should rather add section somewhere during the steps to clear cache here using symfony cache pool command
- OK: 2.5.3 page builder is covered in instructions here
- OK: 2.5.6 Password expiration should be covered by `dbupdate-6.13.0-to-7.5.7`

!!! note

    Version of the migration script in ezplatform-xmltext-fieldtype prior to v1.6.0 would fail to convert embedded images correctly. If you have a database which you have already converted with an old version, you may rerun the `convert-to-richtext` command with the following options:

    `php bin/console ezxmltext:convert-to-richtext --fix-embedded-images-only -v`
    There is no corresponding `ImagesClassList[]` setting in eZ Platform, images is rather an editorial choice. So even though you have custom image classes, you don't need to configure this in the eZ Platform configuration when migrating.

>     The use of `--image-content-types` is also supported together with `--fix-embedded-images-only`. Use it to specify custom image Content Types.

Author: I wanted to check with you on this, as the part below and this seems to better fit as a `**Fixing wrong image type afterwards**` section.

Given `--fix-embedded-images-only` seems to be about being able to fix this afterwards and `--image-content-types=` is already covered just above here for the migration itself now.

So all in all I tried to fix/simplify something that I thought was bug caused by several iterations done on this over the years.

>     The use of `--image-content-types` is also supported together with `--fix-embedded-images-only`. Use it to specify custom image Content Types.

Author: Hence the fix on section below to make sure it correctly describes how to re run migration for updating image embeds.

>     The use of `--image-content-types` is also supported together with `--fix-embedded-images-only`. Use it to specify custom image Content Types.

Reviewer 2: ok. I see

| `cssborder`   | Not supported |   Use `string` as workaround |

###### 3.2.2 Migrate eZ Matrix

**If** you use Matrix field (ezmatrix), you'll need to [migrate the storage format](../updating/4_update_2.5.md#changes-to-matrix-field-type)
as it has changed to json internally for new eZ Platform field type.

###### 3.2.3 Add other eZ Enterprise schemas (eZ Enterprise only)

In eZ Platform, the system ships with additional enterprise features that need to be installed.

1. First add all tables needed by eZ Platform Enterprise:

>    - https://github.com/ezsystems/ezplatform-ee-installer/blob/2.4/Resources/sql/schema.sql#L198

Reviewer 2: Sure we can start on line 198 and skip everything before?

When I do upgrades, I [remove all the "drop table ..." statements in the file](https://github.com/vidarl/ezp-toolkit/blob/master/database/upgrade_db.sh#L25) and then apply the whole thing. I don't think any tables in that file will exists if upgrading from 5.4

Author: ```suggestion
   - https://github.com/ezsystems/ezplatform-ee-installer/blob/2.4/Resources/sql/schema.sql
```
